com.unity.render-pipelines.universal/Runtime/Decal/Entities/DecalCreateDrawCallSystem.cs

```diff
@@ -0,0 +1,203 @@
+using Unity.Collections;
+using Unity.Jobs;
+using Unity.Mathematics;
+
+namespace UnityEngine.Rendering.Universal
+{
+    internal struct DecalSubDrawCall
+    {
+        public int start;
+        public int end;
+        public int count { get => end - start; }
+    }
+
+    internal class DecalDrawCallChunk : DecalChunk
+    {
+        public NativeArray<float4x4> decalToWorlds;
+        public NativeArray<float4x4> normalToDecals;
+        public NativeArray<DecalSubDrawCall> subCalls;
+        public NativeArray<int> subCallCounts;
+
+        public int subCallCount { set { subCallCounts[0] = value; } get => subCallCounts[0]; }
+
+        public override void Push()
+        {
+            count++;
+        }
+
+        public override void RemoveAtSwapBack(int entityIndex)
+        {
+            RemoveAtSwapBack(ref decalToWorlds, entityIndex, count);
+            RemoveAtSwapBack(ref normalToDecals, entityIndex, count);
+            RemoveAtSwapBack(ref subCalls, entityIndex, count);
+            count--;
+        }
+
+        public override void SetCapacity(int newCapacity)
+        {
+            ResizeNativeArray(ref decalToWorlds, newCapacity);
+            ResizeNativeArray(ref normalToDecals, newCapacity);
+            ResizeNativeArray(ref subCalls, newCapacity);
+            capacity = newCapacity;
+        }
+
+        public override void Dispose()
+        {
+            subCallCounts.Dispose();
+
+            if (capacity == 0)
+                return;
+
+            decalToWorlds.Dispose();
+            normalToDecals.Dispose();
+            subCalls.Dispose();
+            count = 0;
+            capacity = 0;
+        }
+    }
+
+    internal class DecalCreateDrawCallSystem
+    {
+        private DecalEntityManager m_EntityManager;
+        private ProfilingSampler m_Sampler;
+
+        public DecalCreateDrawCallSystem(DecalEntityManager entityManager)
+        {
+            m_EntityManager = entityManager;
+            m_Sampler = new ProfilingSampler("DecalCreateDrawCallSystem.Execute");
+        }
+
+        public void Execute()
+        {
+            using (new ProfilingScope(null, m_Sampler))
+            {
+                for (int i = 0; i < m_EntityManager.chunkCount; ++i)
+                    Execute(m_EntityManager.cachedChunks[i], m_EntityManager.culledChunks[i], m_EntityManager.drawCallChunks[i], m_EntityManager.cachedChunks[i].count);
+            }
+        }
+
+        private void Execute(DecalCachedChunk cachedChunk, DecalCulledChunk culledChunk, DecalDrawCallChunk drawCallChunk, int count)
+        {
+            if (count == 0)
+                return;
+
+            DrawCallJob drawCallJob = new DrawCallJob()
```

Comment:
We would probably benefit from smaller jobs so we can go wider. But this would be more of a future improvement.